src/Model/GContact.php

```diff
@@ -0,0 +1,1004 @@
+			&& PortableContact::reachable($gcontact['url'], $gcontact['server_url'], $gcontact['network'], false)
+		) {
+			$data = Probe::uri($gcontact['url']);
+	
+			if ($data["network"] == NETWORK_PHANTOM) {
+				throw new Exception('Probing for URL '.$gcontact['url'].' failed');
+			}
+	
+			$orig_profile = $gcontact['url'];
+	
+			$gcontact["server_url"] = $data["baseurl"];
+	
+			$gcontact = array_merge($gcontact, $data);
+	
+			if ($alternate && ($gcontact['network'] == NETWORK_OSTATUS)) {
+				// Delete the old entry - if it exists
+				$r = q("SELECT `id` FROM `gcontact` WHERE `nurl` = '%s'", dbesc(normalise_link($orig_profile)));
+				if (DBM::is_result($r)) {
+					q("DELETE FROM `gcontact` WHERE `nurl` = '%s'", dbesc(normalise_link($orig_profile)));
+					q("DELETE FROM `glink` WHERE `gcid` = %d", intval($r[0]["id"]));
+				}
+			}
+		}
+	
+		if (!isset($gcontact['name']) || !isset($gcontact['photo'])) {
+			throw new Exception('No name and photo for URL '.$gcontact['url']);
+		}
+	
+		if (!in_array($gcontact['network'], array(NETWORK_DFRN, NETWORK_OSTATUS, NETWORK_DIASPORA))) {
+			throw new Exception('No federated network ('.$gcontact['network'].') detected for URL '.$gcontact['url']);
+		}
+	
+		if (!isset($gcontact['server_url'])) {
+			// We check the server url to be sure that it is a real one
+			$server_url = PortableContact::detectServer($gcontact['url']);
+	
+			// We are now sure that it is a correct URL. So we use it in the future
+			if ($server_url != "") {
+				$gcontact['server_url'] = $server_url;
+			}
+		}
+	
+		// The server URL doesn't seem to be valid, so we don't store it.
+		if (!PortableContact::checkServer($gcontact['server_url'], $gcontact['network'])) {
+			$gcontact['server_url'] = "";
+		}
+	
+		return $gcontact;
+	}
+
+	/**
+	 * @brief Link the gcontact entry with user, contact and global contact
+	 *
+	 * @param integer $gcid Global contact ID
+	 * @param integer $uid  User ID
+	 * @param integer $cid  Contact ID
+	 * @param integer $zcid Global Contact ID
+	 *
+	 */
+	public static function link($gcid, $uid = 0, $cid = 0, $zcid = 0)
+	{
+		if ($gcid <= 0) {
+			return;
+		}
+	
+		$r = q(
+			"SELECT * FROM `glink` WHERE `cid` = %d AND `uid` = %d AND `gcid` = %d AND `zcid` = %d LIMIT 1",
+			intval($cid),
+			intval($uid),
+			intval($gcid),
+			intval($zcid)
+		);
+	
+		if (!DBM::is_result($r)) {
+			q(
+				"INSERT INTO `glink` (`cid`, `uid`, `gcid`, `zcid`, `updated`) VALUES (%d, %d, %d, %d, '%s') ",
+				intval($cid),
+				intval($uid),
+				intval($gcid),
+				intval($zcid),
+				dbesc(datetime_convert())
+			);
+		} else {
+			q(
+				"UPDATE `glink` SET `updated` = '%s' WHERE `cid` = %d AND `uid` = %d AND `gcid` = %d AND `zcid` = %d",
+				dbesc(datetime_convert()),
+				intval($cid),
+				intval($uid),
+				intval($gcid),
+				intval($zcid)
+			);
+		}
+	}
+
+	public static function alternateOstatusUrl($url)
```

Comment:
I believe this should be in `PortableContact` as `alternateOStatusURL()` to honor acronyms.

Reply:
This was not a poco function, so I did not move it. I will move it if you like. Please confirm 😆

Reply:
oops.. nevermind on that

Reply:
Yeah, it's okay, I saw it late as well.

Reply:
done